Expand skip spaces

fernape opened the pull request:
Expand `skipSpaces` so it handles new lines and tabs. This will be useful since queries in code tend to be indented for clarity. Since it now handles more than spaces, rename the function to `skipBlanks`

Aflynn50 committed:
Add /r and refactor

TheSignPainter98 commented on internal/parse/parser_internal_test.go:
```diff
@@ -29,11 +29,12 @@ func TestRunTable(t *testing.T) {
+		{stringf0: p.skipBlanks, result: []bool{true}, input: "     abcd", data: []string{}},
+		{stringf0: p.skipBlanks, result: []bool{true}, input: "  \t  abcd", data: []string{}},
+		{stringf0: p.skipBlanks, result: []bool{true}, input: "\t  abcd", data: []string{}},
+		{stringf0: p.skipBlanks, result: []bool{true}, input: "\n\t  abcd", data: []string{}},
```
There's no test for `\r` here :)

fernape replied:
Done.

Thanks!

TheSignPainter98 commented on internal/parse/parser.go:
```diff
@@ -177,15 +177,18 @@ func (p *Parser) skipByteFind(b byte) bool {
+// skipBlanks advances the parser jumping over consecutive spaces, tabs or new
+// lines. It stops when finding a non-blank character. Returns true if the
+// parser position was actually changed, false otherwise.
+func (p *Parser) skipBlanks() bool {
```
The middle sentence seems redundant to me. Perhaps something like this?
```go
// skipBlanks advances the parser past spaces, tabs and newlines. Returns whether
// the parser position was changed.
```

fernape replied:
Simpler and clearer. Thanks!

fernape committed:
Add more tests.

Include missing test for \r (kcza)

fernape committed:
Simplify comment.

solmyrcan commented on internal/parse/parser_internal_test.go:
```diff
@@ -29,11 +29,16 @@ func TestRunTable(t *testing.T) {
-		{stringf0: p.skipSpaces, result: []bool{true}, input: "  \t  abcd", data: []string{}},
-		{stringf0: p.skipSpaces, result: []bool{false}, input: "\t  abcd", data: []string{}},
+		{stringf0: p.skipBlanks, result: []bool{false}, input: "", data: []string{}},
+		{stringf0: p.skipBlanks, result: []bool{false}, input: "abc    d", data: []string{}},
```
Would it be relevant to have test cases for `\n \r \t` also mixed up in the whitespaces as in here:
`input: "abc  \t  d"`
Just used `\t` here, but maybe others as well?

fernape replied:
Those should be covered with the tests around line 37 and such.

Aflynn50 committed:
Rename skipSpaces to skipBlanks and expand logic.

Jump over spaces, tabs and new lines. This will be necessary since it is
easy to have a query embedded in code and indented with those
characters.

Since it now skips more than only spaces, change the name of the
function.

niemeyer commented on internal/parse/parser.go:
```diff
@@ -177,15 +177,17 @@ func (p *Parser) skipByteFind(b byte) bool {
+		'\r': true,
+		'\n': true,
+	}
+	for p.pos < len(p.input) && blanks[p.input[p.pos]] {
```
This is a longer and much more expensive version of this:
```
for ... {
    switch p.input[p.pos] {
    case ' ', '\t', ...:
        ...
    }
}
```

fernape replied:
Done. Thanks.

niemeyer reviewed (changes requested):
Idea is good, but the implementation could be tuned a bit per notes below.

fernape committed:
Use more performant loop.
